SI-9206 Fix REPL code indentation

som-snytt opened the pull request:
To make code in error messages line up with the original line of
code, templated code is indented by the width of the prompt.

Use the raw prompt (without ANSI escapes or newlines) to determine
the indentation.

Also, indent only once per line.

som-snytt committed:
SI-9206 BooleanProp if set and not untrue

Previously, handy `sys.BooleanProp.keyExists` ignored the
property value. While trying not to make any real estate puns,
this commit will let it go false if a value is supplied that
is not true in the usual Java sense. But what is truth?

Allows `scala -Dscala.color=off`, for example.

som-snytt committed:
SI-9206 REPL prompt is more easily configured

The scala shell prompt can be provided as either a system
property or in compiler.properties.

The prompt string is taken as a format string with one
argument that is the version string.

```
$ scala -Dscala.repl.prompt="%nScala %s> "
Welcome to Scala version 2.11.7-20150616-093756-43a56fb5a1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_45).
Type in expressions to have them evaluated.
Type :help for more information.

Scala 2.11.7-20150616-093756-43a56fb5a1> 42
res0: Int = 42

Scala 2.11.7-20150616-093756-43a56fb5a1> :quit
```

som-snytt committed:
SI-9206 Fix REPL code indentation

To make code in error messages line up with the original line of
code, templated code is indented by the width of the prompt.

Use the raw prompt (without ANSI escapes or newlines) to determine
the indentation.

Also, indent only once per line.

som-snytt committed:
SI-9206 Verbose REPL prompt for info mode

Only for exactly `-Dscala.repl.info`, include the complete version
number string in the REPL prompt. One could imagine this is the mode
for posting snippets to stackoverflow.

som-snytt commented:
One test got by me because --update-check will accidentally make a check file for a SessionTest, which has the checkfile as it were embedded in the session transcript.

som-snytt commented:
/rebuild

Hoping it's one of those spurious IDE build failures.

som-snytt commented:
I guess not. I'd hate to bug him, but I wonder if @adriaanm knows if something around presentation compiler and OSGI has changed.

```
2015-06-18 23:52:33,531 DEBUG [main] - PresentationCompilerActivityListener - Starting PresentationCompilerActivityListener for project expr-eval-code-completion
2015-06-18 23:52:34,223 DEBUG [main] - SimpleContentProposalProvider$ - Cannot load returned type for someNull
com.sun.jdi.ClassNotLoadedException: Type has not been loaded
[snip]
2015-06-18 23:52:34,498  INFO [main] - SimpleContentProposalProviderIntegrationTest$ - Test SimpleContentProposalProviderIntegrationTest finished
2015-06-18 23:52:34,514  INFO [main] - ScalaProject - shutting down compilers for expr-eval-code-completion
2015-06-18 23:52:34,515 DEBUG [main] - PresentationCompilerActivityListener - Stopping PresentationCompilerActivityListener for project expr-eval-code-completion
2015-06-18 23:52:34,515  INFO [main] - ScalaPresentationCompiler - shutting down presentation compiler on project: expr-eval-code-completion
Tests run: 584, Failures: 0, Errors: 1, Skipped: 27, Time elapsed: 1,198.32 sec <<< FAILURE!

Results :

Tests in error: 
  javaVarArgConstructor(org.scalaide.debug.internal.expression.features.NewKeywordTest): Compilation failed

Tests run: 584, Failures: 0, Errors: 1, Skipped: 27
```

Maybe I'll just sneak by his office and see if the door is open. If he has a door.

adriaanm commented:
Who needs doors if they have headphones. Try rebasing? I think we just merged a tandem PR with scalaide related to this, but happy hour may be playing tricks on me.

retronym commented:
Yep, rebasing will fix that one.

som-snytt commented:
This is the rebased result. Is there a command to issue like rebuild or sync? Sorry I'm out of the loop.

retronym commented:
A test failed in [validate-test@bb361e7](https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/902/console)

``` diff
% diff /home/jenkins/workspace/scala-2.11.x-validate-test/test/files/jvm/interpreter-jvm.log /home/jenkins/workspace/scala-2.11.x-validate-test/test/files/jvm/interpreter.check
@@ -366,5 +366,5 @@ plusOne: (x: Int)Int
 res0: Int = 6
 res0: String = after reset
 <console>:8: error: not found: value plusOne
-                plusOne(5) // should be undefined now
-                ^
+              plusOne(5) // should be undefined now
+              ^

```

retronym commented:
Scabot answers to [these commands](https://github.com/scala/scabot#commands).

som-snytt commented:
Thanks, Jason, now I see that I must check build history in Jenkins. I was dissuaded by the lack of links here.

I moved changing the default \nscala> to %nscala> to the next commit, that actually fixes indentation fix-up. The old assumption was to drop one char only from prompt; that test did manual intp.reset(), the .check was two chars off because of double indent; but it's not obvious to me why the output was different for only that one test case. TODO, perhaps. Also, hash tag things that shouldn't be hard to reason about.

SethTisue commented:
@adriaanm, should we go ahead and merge this for 2.11.7, or is it better to leave it alone for now given that REPL stuff is also getting torn up in #4563?

som-snytt commented:
The checks are always greener on the other side.

adriaanm commented:
They look pretty green from my desk, but I'm struggling to get 2.11.7 out the door in time as it is (to make matters worse the elevators are stopping on random floors today).

adriaanm commented:
Hope you don't mind my reopening and moving back to 2.11.7. I didn't make my deadline yesterday.

som-snytt commented:
@adriaanm I was just fixing pasting with a custom prompt. About to push it.

som-snytt commented:
So the reason we need `r"a\|b"` is because `s"a\|b".r` is borked because of the escape. That is, `s"a\\|b".r` for literal vbar. There's cognitive dissonance a/k/a confusion in dealing with two processors of escapes. Or make that, a\k\a. Sorry, a\\k\\a.

som-snytt committed:
SI-9206 Accept paste with custom prompt

But sans test.

som-snytt commented:
The intention of the previous code was to obviate the ctl-D by finishing a pasted transcript with a prompt. But there's a bug if there's more than one such empty line. EDIT: The bug was not detecting `scala>` by itself as a lone prompt. If you intend `scala.>` then you must omit the trailing space. Somewhere this commit fixes a trim.

Each empty line initiates replay, but next prompt is taken as an expression. Here res14 - res15 are pasted:

```
scala> 1
res14: Int = 1

scala> 

scala> 

scala> 2
res15: Int = 2

scala> scala> 1

// Detected repl transcript paste: ctrl-D to finish.

res14: Int = 1

scala> 
// Replaying 1 commands from transcript.

scala> 1
res16: Int = 1


scala> 

scala> scala> 
<console>:8: error: object > is not a member of package scala
              scala>
                   ^

scala> 

scala> scala> 2

// Detected repl transcript paste: ctrl-D to finish.

res15: Int = 2
// Replaying 1 commands from transcript.

scala> 2
res18: Int = 2
```

som-snytt committed:
SI-9206 Local refactor to save eyesight

We talk about bit rot but not about how dust accumulates on
code that hasn't been swept since the last time the furniture
was moved around.

adriaanm commented on src/library/scala/sys/BooleanProp.scala:
```diff
@@ -63,12 +63,13 @@ object BooleanProp {
   /** As an alternative, this method creates a BooleanProp which is true
-   *  if the key exists in the map.  This way -Dfoo.bar is enough to be
-   *  considered true.
+   *  if the key exists in the map and is not assigned a value other than "true",
```
The double negation could do with some De Morgan. `!(assigned !"true")` --> `The key exists in the map and is either the empty string or "true"`.

adriaanm replied:
That said, at least the implementation now matches the docs.

som-snytt replied:
You have to read the comment in a mirror. It's written from the standpoint of a falsifier, kind of like what do you ask a Cretan liar. How do I turn it off? Also, everyone knows that Father's Day is also Opposite Day.

adriaanm commented:
LGTM, with suggestions that hopefully will, erm, prompt a follow-up PR.
